### ISSUE TITLE
Understanding the ground plane representation

### ISSUE THREAD

**Issue by maxvandenhoven** [YihongSun/Dynamo-Depth#5]
Hi @YihongSun,

Thanks for the amazing work on Dynamo-Depth!

I was wondering if you could provide some more details on how the ground plane is represented. I would like to use the ground plane estimation class to find the height of the camera above the road, but I don't quite understand what the returned `plane_param` values are supposed to represent. Any help/pointers on this would be greatly appreciated!

**Comment by YihongSun** [comment]
Hello,

Thank you for your interest in our work!

Consider `input_points`  the set of points that are used for estimating the ground plane with the shape (1, N, 3), where each point contains the associated x, y, z values in camera coordinates. Intuitively, the ground plane for an up-right camera would have its normal vector most aligned with the y axis (which is reflected in `vertical_axis=1` [here](https://github.com/YihongSun/Dynamo-Depth/blob/ec886d5c2cf10daa3d7bcdfd3e842c4d80373802/tools.py#L77))

Therefore, for the sake of numerical stability, we model the ground plane as y = ax + bz + c for evaluating the plane fit. So, the three values in `plane_param` represents `[[[a], [b], [c]]]`.

Please refer to the following code snippet for more details.

```python
import torch
from tools import GroundPlane

gplane = GroundPlane(num_points_per_it=5, max_it=100, tol=0.005)

# Generate random points on the ground plane in camera coordinate via Monte Carlo
num_total = 1000 # number of points sampled on the ground plane
cam_h = 1        # height of camera
noise = 0.01     # vertical noise of the points
radius = 10      # range of sampling in the x and z direction

rand_X = (torch.rand(num_total) - 0.5) * 2 * radius
rand_Y = cam_h+(torch.rand(num_total) - 0.5) * 2 * noise
rand_Z = torch.rand(num_total) * 2 * torch.rand(num_total)
input_points = torch.stack((rand_X,rand_Y, rand_Z)).permute(1,0)[None] # (1, 1000, 3)
print(f'X range: [{torch.min(rand_X).item():.2f}, {torch.max(rand_X).item():.2f}]')
print(f'Y range: [{torch.min(rand_Y).item():.2f}, {torch.max(rand_Y).item():.2f}]')
print(f'Z range: [{torch.min(rand_Z).item():.2f}, {torch.max(rand_Z).item():.2f}]\n')

# Estimate ground plane y = ax + bz + c where plane_param = [[[a], [b], [c]]]
plane_param = gplane.estimate_ground_plane(input_points)
print(plane_param)

# Predict y value from given xz1 input
x_z_1 = torch.cat((rand_X[:,None], rand_Z[:,None], torch.ones(num_total, 1)),1)
pred_y = x_z_1 @ plane_param
print('\nAvg L1 Loss:', torch.abs(pred_y - rand_Y).mean().item())
```
Output:
```text
X range: [-10.00, 10.00]
Y range: [0.99, 1.01]
Z range: [0.00, 1.88]

tensor([[[2.0904e-04],
         [1.4779e-03],
         [9.9991e-01]]])

Avg L1 Loss: 0.005116328597068787
```
Feel free to reply if you have any other questions or concerns!

**Comment by maxvandenhoven** [comment]
Thanks for the clear explanation, this is exactly what I was looking for!

Just to be sure, does the implementation of RANSAC in the `GroundPlane.estimate_ground_plane()` method differ from the classical algorithm in any way? It seems to correspond quite well to the [pseudocode on Wikipedia](https://en.wikipedia.org/wiki/Random_sample_consensus#:~:text=The%20generic%20RANSAC%20algorithm%20works%20as%20the%20following), but I am curious if a trade-off was made to parallelize the algorithm.

**Comment by YihongSun** [comment]
The implemented parallelization should not effect prediction, since the iterative task is quite parallelizable.

I do would like to note that the `d` parameter is omitted for the sake of simplicity, as we select the best fit plane regardless of how good the actual fit is. If needed, simply comparing the best fit with `d` at the end of the algorithm would be sufficient.

`d – The number of close data points (inliers) required to assert that the model fits well to the data.`
